torch/utils/data/datasets/decoder.py

```diff
@@ -0,0 +1,270 @@
+
+                result = np.asarray(img)
+                assert result.dtype == np.uint8, "numpy image array should be type uint8, but got {}".format(result.dtype)
+
+                if etype == "uint8":
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result)
+                else:
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result) / 255.0
+            return None
+
+def imagehandler(imagespec):
+    return ImageHandler(imagespec)
+
+
+################################################################
+# torch video
+################################################################
+
+
+def torch_video(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in "mp4 ogv mjpeg avi mov h264 mpg webm wmv".split():
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchvision.io  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchvision.io.read_video(fname)
+
+
+################################################################
+# torchaudio
+################################################################
+
+
+def torch_audio(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in ["flac", "mp3", "sox", "wav", "m4a", "ogg", "wma"]:
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchaudio  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchaudio.load(fname)
+
+
+
+################################################################
+# a sample decoder
+################################################################
+
+
+class Decoder:
+    """
+    Decode key/data sets using a list of handlers.
+    For each key/data item, this iterates through the list of
+    handlers until some handler returns something other than None.
+    """
+
+    def __init__(self, handlers):
+        self.handlers = handlers
+
+    def add_decoder(self, decoder):
+        if not decoder:
+            return
+        if not self.handlers:
+            self.handlers = [decoder]
+        else:
+            self.handlers.append(decoder)
+
+    def decode1(self, key, data):
+        if not data:
+            return data
+
+        # if data is a stream handle, we need to read all the content before decoding
+        if isinstance(data, io.BufferedIOBase) or isinstance(data, io.RawIOBase):
+            data = data.read()
+
+        for f in self.handlers:
+            result = f(key, data)
+            if result is not None:
+                return result
```

Comment:
Why we do iteration over all handlers? Several problems:
1. Hard to overwrite some extension handler.
2. Overhead for iteration over all handlers, and iteration over all extension types. (each data goes to all handlers until find corresponding)
3. Not find handler -> return None

Reply:
These codes are adopted from Nvidia, we can change it. 
1. They already have a helper function to overwrite extension handler (`handle_extension`).  
2. Iteration does have overhead, probably we can switch to dict or other types. 
3. We probably can raise error here.